Controller/Giftcard/AddGiftCard.php

```diff
@@ -0,0 +1,225 @@
+
+        if (!$giftCardCode) {
+            return $this->jsonResponse(false, __('Gift Card code is required.'));
+        }
+
+        try {
+            $quote = $this->checkoutSession->getQuote();
+
+            // Validate quote
+            if (!$quote || !is_object($quote)) {
+                return $this->jsonResponse(false, __('We couldn\'t retrieve your cart details. Please try again.'));
+            }
+
+            if ($isCartPage && !$quote->isVirtual()) {
+                $this->updateShippingMethod($quote, $selectedShippingMethod);
+            }
+
+            if ($this->isCartTotalZero($quote)) {
+                return $this->jsonResponse(false, __('Your cart total is zero, so there\'s no need to apply a Gift Card code.'));
+            }
+
+            $apiResponse = $this->giftcardHelper->checkGiftCardBalanceStatus($giftCardCode);
+
+            return $this->handleApiResponse($apiResponse, $giftCardCode, $quote);
+
+        } catch (LocalizedException $e) {
+            return $this->jsonResponse(false, $e->getMessage());
+        }
+    }
+
+    /**
+     * Update the shipping method of the quote
+     *
+     * @param mixed $quote
+     * @param mixed $selectedShippingMethod
+     * @return void
+     */
+    private function updateShippingMethod($quote, $selectedShippingMethod)
+    {
+        $shippingMethod = $quote->getShippingAddress()->getShippingMethod();
+        $shippingAmount = $quote->getShippingAddress()->getShippingAmount();
+
+        if ($selectedShippingMethod &&
+            ($shippingMethod !== $selectedShippingMethod || $shippingAmount == 0)) {
+            $this->giftcardHelper->setShippingMethodForce(
+                $quote,
+                $selectedShippingMethod
+            );
+        }
+    }
+
+    /**
+     * Check if the cart total is zero
+     *
+     * @param mixed $quote
+     * @return bool
+     */
+    private function isCartTotalZero($quote)
+    {
+        return (float)$quote->getGrandTotal() === 0.0 && $quote->getShippingAddress()->getShippingAmount() == 0;
+    }
+
+    /**
+     * Handle the API response for gift card validation.
+     *
+     * @param mixed $apiResponse
+     * @param mixed $giftCardCode
+     * @param mixed $quote
+     * @return mixed
+     */
+    private function handleApiResponse($apiResponse, $giftCardCode, $quote)
+    {
+        $apiErrorMessage = $apiResponse['xError'] ?? null;
+        $status = $apiResponse['xStatus'] ?? null;
+        $activationStatus = $apiResponse['xActivationStatus'] ?? null;
+        $remainingBalance = $apiResponse['xRemainingBalance'] ?? null;
+
+        if ($status === "Approved" && $activationStatus === "Active" && $remainingBalance > 0) {
+            return $this->applyGiftCard($apiResponse, $giftCardCode, $quote);
+        }
+
+        if ($activationStatus === 'Inactive') {
+            return $this->jsonResponse(false, __('Your gift card account is inactive. Please activate it before proceeding.'));
```

Comment:
Is this message shown to the customer or only admin?
Do customers have the ability to activate gift cards?

Reply:
@mstein-ck 
**1. Is this message shown to the customer or only admin?
Comment:** This message shows to the customer while adding giftcard.

**2. Do customers have the ability to activate gift cards?
Comment:** No, Customers do not have the ability to activate gift cards as Yaakov told we do not provide that functionality to customer right now.

Reply:
In that case can you update the message to remove the suggestion to activate the card? Same for 0 dollar balance below

Could all customer facing messages be placed in a separate file for easier review?
```suggestion
            return $this->jsonResponse(false, __('Your gift card account is inactive'));
```

Reply:
suggestion applied.